src/SQL/Parser.php

```diff
@@ -79,40 +82,28 @@ public function __construct(bool $mySQLStringEscaping)
      */
     public function parse(string $sql, Visitor $visitor): void
     {
-        /** @var array<string,callable> $patterns */
-        $patterns = [
-            self::NAMED_PARAMETER => static function (string $sql) use ($visitor): void {
-                $visitor->acceptNamedParameter($sql);
-            },
-            self::POSITIONAL_PARAMETER => static function (string $sql) use ($visitor): void {
-                $visitor->acceptPositionalParameter($sql);
-            },
-            $this->sqlPattern => static function (string $sql) use ($visitor): void {
-                $visitor->acceptOther($sql);
-            },
-            self::SPECIAL => static function (string $sql) use ($visitor): void {
-                $visitor->acceptOther($sql);
-            },
-        ];
-
         $offset = 0;
-
-        while (($handler = current($patterns)) !== false) {
-            if (preg_match('~\G' . key($patterns) . '~s', $sql, $matches, 0, $offset) === 1) {
-                $handler($matches[0]);
-                reset($patterns);
-
-                $offset += strlen($matches[0]);
+        $length = strlen($sql);
+        while ($offset < $length) {
+            if (preg_match($this->tokenPattern, $sql, $matches, 0, $offset) === 1) {
+                $match = $matches[0];
+                if (isset($matches['named']) && strlen($matches['named']) > 0) {
+                    $visitor->acceptNamedParameter($match);
+                } elseif (isset($matches['pos']) && strlen($matches['pos']) > 0) {
+                    $visitor->acceptPositionalParameter($match);
+                } else {
+                    $visitor->acceptOther($match);
+                }
+
+                $offset += strlen($match);
             } elseif (preg_last_error() !== PREG_NO_ERROR) {
                 // @codeCoverageIgnoreStart
                 throw RegularExpressionError::new();
                 // @codeCoverageIgnoreEnd
             } else {
-                next($patterns);
+                throw new RuntimeException(sprintf('Unexpected token near "%s"', substr($sql, $offset, 10)));
```

Comment:
When will this happen? Please add a test. What does 10 mean in the `substr()` call?